pennylane/gradients/finite_difference.py

```diff
@@ -0,0 +1,258 @@
+
+    if form == "forward":
+        shifts = np.arange(N, dtype=np.float64)
+
+    elif form == "backward":
+        shifts = np.arange(-N + 1, 1, dtype=np.float64)
+
+    elif form == "center":
+        if order % 2 != 0:
+            raise ValueError("Centered finite-difference requires an even order.")
+
+        N = num_points // 2
+        shifts = np.arange(-N, N + 1, dtype=np.float64)
+
+    else:
+        raise ValueError(f"Unknown form {form}. Must be one of 'forward', 'backward', 'center'.")
+
+    A = shifts ** np.arange(len(shifts)).reshape(-1, 1)
+    b = np.zeros_like(shifts)
+    b[n] = factorial(n)
+    coeffs = np.linalg.solve(A, b)
+
+    stencil = np.stack([coeffs, shifts])
+    stencil[0, np.abs(stencil[0, :]) < 1e-10] = 0
+    stencil = stencil[:, ~np.all(stencil == 0, axis=0)]
+    stencil = stencil[:, np.argsort(np.abs(stencil)[1])]
+    return stencil
+
+
+def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
+    r"""Generate a list of tapes where the corresponding trainable parameter
+    index has been shifted by the values given.
+
+    Args:
+        tape (.QuantumTape): input quantum tape
+        idx (int): trainable parameter index to shift the parameter of
+        shifts (Sequence[float or int]): sequence of shift values
+        multipliers (Sequence[float or int]): Sequence of multiplier values to
+            scale the paraameter by. If not provided, the parameter will
+            not be scaled.
+
+    Returns:
+        list[QuantumTape]: List of quantum tapes. Each tape has parameter
+        ``idx`` shifted by consecutive values of ``shift``. The length
+        of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = qml.math.stack(tape.get_parameters())
+    tapes = []
+
+    for i, s in enumerate(shifts):
+        shifted_tape = tape.copy(copy_operations=True)
+
+        shift = np.zeros(qml.math.shape(params), dtype=np.float64)
+        shift[idx] = s
+
+        a = multipliers[i] if multipliers is not None else 1.0
+
+        shifted_params = a * params + qml.math.convert_like(shift, params)
+        shifted_tape.set_parameters(qml.math.unstack(shifted_params))
+
+        tapes.append(shifted_tape)
+
+    return tapes
+
+
+def finite_diff(tape, argnum=None, h=1e-7, order=1, n=1, form="forward", f0=None):
+    r"""Generate the parameter-shift tapes and postprocessing methods required
+    to compute the gradient of a gate parameter with respect to an
+    expectation value.
+
+    Args:
+        tape (.QuantumTape): quantum tape to differentiate
+        argnum (int or list[int] or None): Trainable parameter indices to differentiate
+            with respect to. If not provided, the derivatives with respect to all
+            trainable indices are returned.
+        h (float): finite difference method step size
+        order (int): The approximation order of the finite difference method to use.
+        n (int): compute the :math:`n`-th derivative

```

Comment:
I just [see](https://en.wikipedia.org/wiki/Second_derivative) that n-th derivative and n-th order derivative are used interchangably...so maybe ignore my confusion about the term above.

Reply:
No, it's a good catch, I should make the language more explicit.